src/lib.rs

```diff
@@ -16,168 +17,256 @@ use std::{
         unsafe {
-            let me = compiler.funcs.IRMetalLibBinaryCreate();
-            Ok(Self {
+            let me = NonNull::new(compiler.funcs.IRMetalLibBinaryCreate())
+                .expect("Failed to create empty IRMetalLibBinary");
+            Self {
                 me,
                 funcs: compiler.funcs.clone(),
-            })
+            }
         }
     }
 
-    pub fn get_byte_code(&self) -> Vec<u8> {
-        let size_in_bytes = unsafe { self.funcs.IRMetalLibGetBytecodeSize(self.me) };
-        let mut bytes = vec![0u8; size_in_bytes];
-        let _written = unsafe {
+    #[doc(alias("IRMetalLibGetBytecode", "IRMetalLibGetBytecodeSize"))]
+    pub fn byte_code(&self) -> Vec<u8> {
+        let size_in_bytes = unsafe { self.funcs.IRMetalLibGetBytecodeSize(self.me.as_ptr()) };
+        let mut bytes = Vec::with_capacity(size_in_bytes);
+        let written = unsafe {
             self.funcs
-                .IRMetalLibGetBytecode(self.me, bytes.as_mut_ptr())
+                .IRMetalLibGetBytecode(self.me.as_ptr(), bytes.as_mut_ptr())
         };
+        unsafe { bytes.set_len(written) }
+        assert_eq!(written, size_in_bytes);
         bytes
     }
 }
 
 pub struct IRRootSignature {
-    me: *mut bindings::IRRootSignature,
+    me: NonNull<bindings::IRRootSignature>,
     funcs: Arc<bindings::metal_irconverter>,
 }
 
 impl Drop for IRRootSignature {
+    #[doc(alias = "IRRootSignatureDestroy")]
     fn drop(&mut self) {
-        unsafe { self.funcs.IRRootSignatureDestroy(self.me) }
+        unsafe { self.funcs.IRRootSignatureDestroy(self.me.as_ptr()) }
     }
 }
 
+#[derive(Error, Debug)]
+pub enum RootSignatureError {
+    #[error("Failed to create IRRootSignature: {0:?}")]
+    CreateError(ffi::IRErrorCode),
+    #[error("Failed to create IRRootSignature")]
+    Unknown,
+}
+
 impl IRRootSignature {
+    #[doc(alias = "IRRootSignatureCreateFromDescriptor")]
     pub fn create_from_descriptor(
         compiler: &IRCompiler,
         desc: &ffi::IRVersionedRootSignatureDescriptor,
-    ) -> Result<IRRootSignature, Box<dyn std::error::Error>> {
-        unsafe {
-            let mut error = std::ptr::null_mut();
+    ) -> Result<IRRootSignature, RootSignatureError> {
+        let mut error = std::ptr::null_mut();
 
-            let me = compiler
+        let me = NonNull::new(unsafe {
+            compiler
                 .funcs
-                .IRRootSignatureCreateFromDescriptor(desc, &mut error);
-
-            Ok(Self {
-                me,
-                funcs: compiler.funcs.clone(),
-            })
+                .IRRootSignatureCreateFromDescriptor(desc, &mut error)
+        });
+
+        // If the root signature failed to create
+        if !error.is_null() {
+            // IRErrorCode is #[repr(u32)], so this transmute should be fine
+            let error = IRError::from_ptr(error, compiler.funcs.clone());
+            Err(RootSignatureError::CreateError(error.code()))
```

Comment:
Shouldn't this and anything that uses `IRError` include the payload if it's non-NULL and with a nonzero length?